src/libhirte/log/log.h

```diff
@@ -22,31 +22,63 @@ const char *log_level_to_string(LogLevel l) {
 }
 
 
-typedef void (*LogFn)(LogLevel lvl, const char *msg, const char *data);
+typedef void (*LogFn)(
+                LogLevel lvl,
+                const char *file,
+                const char *line,
+                const char *func,
+                const char *msg,
+                const char *data);
+
+// const char *file, const char *line, const char *func,
+void hirte_log_to_journald_with_location(
+                LogLevel lvl,
+                const char *file,
+                const char *line,
+                const char *func,
+                const char *msg,
+                const char *data) {
+        // file and line are required to contain the field names and func gets its key
+        // automatically set (in contrast to what is written in the documentation).
+        // see https://www.freedesktop.org/software/systemd/man/sd_journal_print.html
+        _cleanup_free_ char *file_info = NULL;
+        _cleanup_free_ char *line_info = NULL;
+        asprintf(&file_info, "CODE_FILE=%s", file);
+        asprintf(&line_info, "CODE_LINE=%s", line);
 
-void hirte_log_to_journald(LogLevel lvl, const char *msg, const char *data) {
         // clang-format off
-        sd_journal_send(
+        sd_journal_send_with_location(
+                file_info, line_info, func,
                 "HIRTE_LOG_LEVEL=%s", log_level_to_string(lvl),
                 "HIRTE_MESSAGE=%s", msg,
                 "HIRTE_DATA=%s", data,
                 NULL);
         // clang-format on
 }
 
-void hirte_log_to_stdout(LogLevel lvl, const char *msg, const char *data) {
+void hirte_log_to_stdout_with_location(
+                LogLevel lvl,
+                const char *file,
+                const char *line,
+                const char *func,
+                const char *msg,
+                const char *data) {
+
         time_t t = time(NULL);
         const size_t timestamp_size = 9;
         char timebuf[timestamp_size];
         timebuf[strftime(timebuf, sizeof(timebuf), "%H:%M:%S", localtime(&t))] = '\0';
 
         // clang-format off
-        printf("%s %s\t{msg: %s, data: %s}\n", 
+        fprintf(stderr, "%s %s\t%s:%s %s\t{msg: %s, data: %s}\n", 
                 timebuf, log_level_to_string(lvl), 
+                file, line, func,
                 msg, data);
         // clang-format on
 }
 
+#define hirte_log_to_stdout hirte_log_to_stdout_with_location
+#define hirte_log_to_journald hirte_log_to_journald_with_location
```

Comment:
Do we need these?

Reply:
Not really. I saw this pattern in [systemds sd-journal.h](https://github.com/systemd/systemd/blob/main/src/systemd/sd-journal.h#L51-L55) and adopted it - although its just a renaming use case here. I remove it.